specs: add 'handle' spec doc (#93)
bluesky-social/atproto-website: content/specs/handle.md

- the overall handle must contain only ASCII characters, and can be at most 253 characters long (in practice, handles may be restricted to a slightly shorter length)
- the overall handle is split in to multiple segments (referred to as “labels” in standards documents), separated by ASCII periods (`.`)
- no proceeding or trailing ASCII periods are allowed, and there must be at least two segments. that is, ”bare” top-level domains are not allowed as handles, even if valid “hostnames” and “DNS names”. and “tailing dot” syntax for DNS names is not allowed for handles
- each segment must have at least 1 and at most 63 characters (not including the periods). the allowed characters are ASCII letters (`a-z`), digits (`0-9`), and hyphens (`-`)
- segments can not start or end with a hyphen.
- the last segment (the “top level domain”) can not start with a numeric digit
- handles are not case-sensitive, and should be normalized to lower-case (that is, normalize ASCII `A-Z` to `a-z`)

To be explicit (the above rules already specify this), no whitespace, null bytes, joining characters, or other ASCII control characters are allowed in the handle, including as prefix/suffix.

> Modern “hostnames” (and thus handles) allow ASCII digits in most positions, with the exception that the last segment (top-level domain, TLD) can not start with a digit.

Reviewer: ```suggestion
Modern “hostnames” (and thus handles) allow ASCII digits in most positions, with the exception that the last segment (top-level domain, TLD) cannot start with a digit.
```

```
jay.bsky.social
8.cn
name.t--t        // not a real TLD, but syntax ok
XX.LCS.MIT.EDU
a.co
xn--notarealidn.com
xn--fiqa61au8b7zsevnm8ak20mc4a87e.xn--fiqs8s
xn--ls8h.test
example.t        // not a real TLD, but syntax ok
```

Invalid syntax:

```
jo@hn.test
💩.test
john..test
xn--bcher-.tld
john.0
cn.8
www.masełkowski.pl.com
org
name.org.
```

Valid syntax, but must always fail resolution due to other restrictions:

```
2gzyxa5ihm7nsggfxnu52rck2vv4rvmdlkiu3zzui5du4xyclen53wid.onion
laptop.local
blah.arpa
```

### Usage and Implementation Guidelines

Handles may be prefixed with the “at” symbol (like `@jay.bsky.team`) in user interfaces, but this is not a valid syntax for a handle in records, APIs, and other back-end contexts.

Internationalized Domain Names (”IDN”, or “punycode”) are not directly relevant to the low-level handle syntax. In their encoded form, IDNs are already valid hostnames, and thus valid handles. Such handles must be stored and transmitted in encoded ASCII form. Handles that “look like” IDNs, but do not parse as valid IDNs, are valid handles, just as they are valid hostnames. Applications may, optionally, parse and display IDN handles as Unicode.

> Handles are not case-sensitive, which means they can be safely normalized from user input to lower-case (ASCII) form. Only normalized (lower-case) handles should be stored in records or used in outbound API calls. Applications should not preserve user-provided case information and attempt to display handles in anything other than lower-case. For example, the handle input string `[BlueskyWeb.xyz](http://www.BlueskyWeb.xyz)`  should be normalized, stored, and displayed as `blueskyweb.xyz`.

Reviewer 2: Should this be a markdown link or just `BlueskyWeb.xyz` ?

Reviewer 2: Also applies to links further down (paypal, jay.bsky.social)

Author: thanks!

Handles should generally not be truncated to local context. For example, the handle `@jay.bsky.social` should not be displayed as `@jay`, even in the local context of a `bsky.social` service.

Providers of handle “namespaces” (eg, as subdomains on a registered domain) may impose any additional limits on handles that they wish. It is recommended to constrain the allowed segment length to something reasonable, and to reserve a common set of segment strings like `www`, `admin`, `mail`, etc. There are multiple public lists of “commonly disallowed usernames” that can be used as a starting point.

From a practical standpoint, handles should be limited to at most 244 characters, fewer than the 253 allowed for DNS names. This is because DNS verification works with the prefix `_atproto.`, which adds 9 characters, and that overall name needs to be valid.

Handle hostnames are expected to be mainstream DNS domain names, registered through the mainstream DNS name system. Handles with non-standard TLDs, or using non-standard naming systems, will fail to interoperate with other network services and protocol implementations in the atproto ecosystem.

> ### Possible Future Changes

Author: That is a real concern. I think we can't realistically retain VariableCase through the entire system, and having inconsistent display of the same identifier raises it's own accessibility and usability issues.

I added to the relevant "usage guidelines" paragraph:

> Long all-lowercase handles can be a readability and accessibility challenge. Sub-domain separation (periods), hyphenation, or use of "display names" in application protocols can all help.

This is kind of a band-aid, but I think it is realistically how people real with this with actual domain names: break.things.up.com or break-things-up.com instead of breakthingsup.com or BreakThingsUp.com
